### DISCUSSION TITLE
Explicit ownership when passing IDisposable

### DISCUSSION THREAD

**Discussion by crhaglun** [dotnet/csharplang#5277] [Language Ideas]
I'm working on a piece of legacy code where we pass around instances of IDisposable classes. Sometimes we let the called method or class dispose the passed instance, sometimes that responsibility remains with the caller.

I'm aware of three main ways of dealing with this:
1. A boolean flag to indicate whether a specific instance should be disposed (following the pattern in the standard library, like one of the [HttpClient constructors](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.-ctor))
2. A wrapper construct where "Dispose()" does not actually dispose; it could be a no-op, some reference counting implementation or other way of preventing direct disposal. [System.Reactive.Disposables in dotnet / reactive](https://github.com/dotnet/reactive/tree/main/Rx.NET/Source/src/System.Reactive/Disposables) has a number of interesting alternatives.
3. Never explicitly dispose and rely on the garbage collection mechanisms to ensure resources are disposed.

Each method has its drawbacks, but one that's common for all is that it's hard to read and reason about changes in pull requests, and unsurprisingly we've had a fair share of disposal bugs.

**Could C# use a fourth way, a language feature to indicate whether "ownership" of disposable instances is transferred or not?**

For want of better terms, I'll try giving some examples using the keyword _borrow_ to indicate "caller is keeping ownership" and _take_ to indicate "caller is giving up ownership".

Keywords when declaring parameters:
````
void SomeMethod(borrow ClassThatImplementsIDisposable instance)
{
    // no 'using', no explicit call to instance.Dispose()
    instance.SomeMethod();
}

void SomeMethod(take ClassThatImplementsIDisposable instance)
{
    // 'using' or explicit call to Dispose() is required
    using (instance)
    {
        SomeMethod(borrow instance);
    }
}
````

When calling methods that require ownership annotations:
````
using (var x = new ClassThatImplementsIDisposable())
{
    SomeMethod(borrow x);
}

SomeMethod(take new ClassThatImplementsIDisposable());
````

**Comment by HaloFour** [comment]
Given that the methods would need to know whether or not they "own" the reference this sounds like something that could be accomplished with attributes and an analyzer.  There wouldn't be a way to do this more dynamically as the `IDisposable` references themselves have no concept of ownership and the one method couldn't behave one way conditionally based on how the method was called.  That could possibly be managed with a wrapped struct that also implements `IDisposable` that internally tracks whether it owns the reference or not, but that wouldn't necessitate a language change for it.

**Reply by crhaglun** [reply]
Thank you, those are good points. Having attributes would allow us to disambiguate who owns an IDisposable when passing an instance to another method or constructor, and a code analyzer could enforce the behavior. 

My main concern here was not our code actually, but external libraries we rely on. But I guess attributes could be useful there too, as long as we can convince the upstreams to pick up the same convention and consistently use such attributes.

Perhaps it's a bit naive, but seeing how the language already has made special accommodations for IDisposable (the 'using' statement) I figured another language-level construct to help determine when we "must" / "must not" use 'using' would be logical :-)

**Comment by ufcpp** [comment]
see also: https://github.com/dotnet/csharplang/discussions/4808 https://github.com/dotnet/csharplang/discussions/859

**Reply by crhaglun** [reply]
Ah, thank you! I also found the still open proposal [dotnet/roslyn #160](https://github.com/dotnet/roslyn/issues/160) which could be a more general solution (i.e. applicable to other types that IDisposable's)

**Comment by jez9999** [comment]
Yeah, I don't get why C# doesn't have this.  I regularly come across the problem of receiving an IDisposable and either 1) not knowing whether I'm supposed to dispose of it or 2) not realizing it even implements IDisposable and the analyzers don't tell me because they can't be sure whether I'm supposed to dispose of it.

It seems to me like the IDisposable interface on its own isn't really fit for purpose and isn't providing enough information.

**Reply by CyrusNajmabadi** [reply]
> Yeah, I don't get why C# doesn't have this.

Ownership is a massively complex feature area in its own right.  See just how long its taken for rust to get here, with an enormous amount of effort and pain that it's taken just to have a model that is still proving out its viability.

**Comment by dlebansais** [comment]
You can use [IDisposableAnalyzers](https://github.com/DotNetAnalyzers/IDisposableAnalyzers/) to analyze the code, and make your own modification in their program to add more support for ownership transfer.

I did something similar to what you ask, adding the `AcquireOwnership` attribute, and a configuration file for 3rd-party libraries (since for them you can't add the attribute). Here is [a link](https://github.com/dlebansais/IDisposableAnalyzers/blob/master/README.md#benefits-of-this-fork).

**Reply by martin-steinhoefer** [reply]
You could also use a counting wrapper around an `IDisposable`, so that **every time** you hand out an instance, you get ownership transfer (of the wrapper). That worked for me (both shared (parallel) and sequential exclusive access to the shared `IDisposable` behind).

**Comment by mikenakis** [comment]
The proper way to handle the situation described by the OP is with good software design, not with a new language feature. What I do for many interfaces is that I have two flavors: one extending `IDisposable` and one not extending `IDisposable`. (What usually happens is that actual interfaces almost never extend `IDisposable`, only classes do.) So, I generally avoid passing around the interfaces that extend `IDisposable`, so the recipient never has to wonder whether they should dispose, because what is handed out to them is not disposable. However, when I do pass something that does extend `IDisposable`, then the caller is always expected to dispose it.  So, the very simple rule that I end up following is "if it is disposable, you must dispose it." Problem solved.

**Reply by jez9999** [reply]
Except for all the stuff built into .NET that implements `IDisposable` and you're not necessarily meant to dispose it, or all the NuGet libs written that do the same.

**Reply by mikenakis** [reply]
Yes, true. For this reason I tend to wrap lots of ready-made stuff in my own classes.

**Reply by dlebansais** [reply]
All these wrappers would not be needed if we could express the intent, with a few attributes, in way similar to `NotNullWhen`.

Food for thoughts.

**Reply by CyrusNajmabadi** [reply]
i would genuinely love if someone protyped this out.  Using normal attributes.  It will help demonstrate the complexity of the space, and what any sort of lang solution would need to be sufficient.

Note: this is why NotNullWhen is not a lang feature, but an attribute.  Trying to model all those relationships with lang syntax is just not a great idea.

**Reply by jez9999** [reply]
@CyrusNajmabadi I'm pretty new to this idea but what would be wrong with simply `Borrows` and `Takes` on the caller side, and `Holds`, `Gives`, and `Shares` on the callee (the latter allowing the caller to decide whether ownership is transferred or not)?  It would be a compile-time error to take from a callee that holds, and to borrow from a callee that gives.

**Reply by dlebansais** [reply]
> i would genuinely love if someone protyped this out. Using normal attributes. It will help demonstrate the complexity of the space, and what any sort of lang solution would need to be sufficient.
> 
> Note: this is why NotNullWhen is not a lang feature, but an attribute. Trying to model all those relationships with lang syntax is just not a great idea.

I don't think anybody here asked for a change in the language beside the OP. We're discussing attributes, which as you say should suffice.

Regarding prototyping the idea, we have IDisposableAnalyzers I linked above, and my fork that defines the attribute, which I'm using, in production code. This covers the ownership transfer request that initiated the discussion, so it's more than prototyped, it's implemented. Is there something else needed? Or do you want to enlarge the discussion to the full analysis of disposable object lifespan?

**Reply by CyrusNajmabadi** [reply]
> but what would be wrong 

I would expect to need a couple of years of use to be anywhere close to being able to answer that.
